### PR TITLE
Group Echo Submission

### PR DESCRIPTION
**Pull request by JYS-99**
Hi Prof. Goodman,

Here's group echo. We would like to make a submission of our Assignment Candidate Generators.

Thanks,
Echo

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -2,3 +2,4 @@ services/db/02_rest.sql filter=lfs diff=lfs merge=lfs -text
 *.sql filter=lfs diff=lfs merge=lfs -text
 *.pkl filter=lfs diff=lfs merge=lfs -text
 * text=auto eol=lf
+services/db/01_user.sql filter=lfs diff=lfs merge=lfs -text
```
Comment (kennethgoodman) [inline .gitattributes:5]: remove this

File: services/backend/src/config.py [hunk 2]
```diff
@@ -8,7 +8,7 @@ class BaseConfig:
     BCRYPT_LOG_ROUNDS = 13
     ACCESS_TOKEN_EXPIRATION = 900  # 15 minutes
     REFRESH_TOKEN_EXPIRATION = 2592000  # 30 days
-    NUMBER_OF_CONTENT_IN_ANN = 100 # UPDATE THIS WHEN DEVELOPING ANN
+    NUMBER_OF_CONTENT_IN_ANN = 10000 # UPDATE THIS WHEN DEVELOPING ANN
```
Comment (kennethgoodman) [inline services/backend/src/config.py:11]: remove this change

File: services/backend/src/data_structures/approximate_nearest_neighbor/two_tower_ann.py [hunk 3]
```diff
@@ -99,7 +99,6 @@ def get_ANN_recommendations_from_user(user_id, team, K):
         user_df = pd.DataFrame(user_engagements)
 
         user_embedding = team_wrappers[team].generate_user_embeddings(user_df)
-
```
Comment (kennethgoodman) [inline services/backend/src/data_structures/approximate_nearest_neighbor/two_tower_ann.py:102]: remove this change

File: services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py [hunk 4]
```diff
@@ -23,13 +25,40 @@ def gather_data(self):
...
+        ).all()
 
     def compute_similarity(self):
         # Compute the similarity between users.
         # For simplicity, we'll use cosine similarity as our metric.
         # Update self.user_similarity_map with the similarities.
-        pass
+        # 1. Create a user-item matrix using dictionaries
+        user_item_dict = {}
+        # filtered_interactions = [interaction for interaction in self.interactions if interaction.engagement_type == EngagementType.Like]
```
Comment (kennethgoodman) [inline services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py:36]: remove comments

File: services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py [hunk 5]
```diff
@@ -23,13 +25,40 @@ def gather_data(self):
...
+        for interaction in self.interactions:
+            content_id, user_id, engagement_type, engagement_value = interaction
+
+            if user_id not in user_item_dict:
+                user_item_dict[user_id] = {}
+            user_item_dict[user_id][content_id] = user_item_dict[user_id].get(content_id, 0) + engagement_value
+        # print(user_item_dict)
```
Comment (kennethgoodman) [inline services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py:43]: remove commented code

File: services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py [hunk 6]
```diff
@@ -23,13 +25,40 @@ def gather_data(self):
...
+        # Convert the dictionary to a list of lists for cosine similarity
+        users = list(user_item_dict.keys())
+        contents = {content_id for user, content_data in user_item_dict.items() for content_id in content_data.keys()}
+        matrix = []
+        for user in users:
+            # print(user)
```
Comment (kennethgoodman) [inline services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py:49]: remove commented code

File: services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py [hunk 7]
```diff
@@ -23,13 +25,40 @@ def gather_data(self):
...
+            row = [user_item_dict[user].get(content_id, 0) for content_id in contents]
+            matrix.append(row)
+        # print(matrix)
+        similarity_matrix = cosine_similarity(matrix)
+        # print(similarity_matrix)
```
Comment (kennethgoodman) [inline services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py:54]: remove commented code

File: services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py [hunk 8]
```diff
@@ -38,4 +67,30 @@ def get_similar_users(self, user_id):
     def recommend_items(self, user_id, num_recommendations=10):
         # For a given user, fetch the list of similar users.
         # Recommend items engaged by those users, which the given user hasn't seen.
-        pass
+        # Fetch the list of similar users for the given user_id.
+        self.compute_similarity()
+        similar_users = sorted(self.user_similarity_map.get(user_id, {}).items(), key=lambda x: x[1], reverse=True)
+        # print(self.user_similarity_map)
```
Comment (kennethgoodman) [inline services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py:73]: remove commented code

File: services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py [hunk 9]
```diff
@@ -38,4 +67,30 @@ def get_similar_users(self, user_id):
...
+        # Track recommended items.
+        recommended_content_ids = set()
+
+        # Items that the user has already engaged with.
+        user_engaged_content = set([content_id for content_id, user1_id, _, _ in self.interactions if user1_id == user_id])
+        # print(user_engaged_content)
+        for similar_user, _ in similar_users:
+            # Get items engaged by the similar user.
+            similar_user_engaged_content = set([content_id for content_id, user_id, _, _ in self.interactions if user_id == similar_user])
+            # print(similar_user_engaged_content)
```
Comment (kennethgoodman) [inline services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py:85]: remove commented code

File: services/backend/src/recommendation_system/recommendation_flow/candidate_generators/echo/YourChoiceGenerator.py [hunk 10]
```diff
@@ -1,9 +1,35 @@
...
+                .offset(offset)
+                .all()
+            )
+            return list(map(lambda x: x[0], results))[:725], ([0]*len(results))[:725]
+        elif starting_point.get("content_id", False):
+            content_ids, scores = ann_with_offset(
+                starting_point["content_id"], 0.9, limit, offset, return_distances=True
+            )
+            return content_ids, scores
+        raise NotImplementedError("Need to provide a key we know about")
```
Comment (kennethgoodman) [inline services/backend/src/recommendation_system/recommendation_flow/candidate_generators/echo/YourChoiceGenerator.py:32]: take this out at this point

File: services/db/01_user.sql [hunk 11]
```diff
@@ -1,18 +1,3 @@
...
-/*!40000 ALTER TABLE `user` ENABLE KEYS */;
-UNLOCK TABLES;
\ No newline at end of file
+version https://git-lfs.github.com/spec/v1
```
Comment (kennethgoodman) [inline services/db/01_user.sql:1]: remove this change

**Review by kennethgoodman — COMMENTED** [review]
please remove the commented code, otherwise looks good to me, I'll let you know if it fails to run in production once you fix

**Comment by JYS-99** [comment]
Hi Prof. Goodman,

We have updated the codes. Please help to test in production. Thanks!

File: .gitattributes (hunk 1 above)
Comment (kennethgoodman) [reply .gitattributes:5]: please remove this

File: df_user.csv
Comment (kennethgoodman) [inline df_user.csv:1]: where is this used? can this be removed?

File: services/backend/src/config.py [hunk 12]
```diff
@@ -8,7 +8,7 @@ class BaseConfig:
     BCRYPT_LOG_ROUNDS = 13
     ACCESS_TOKEN_EXPIRATION = 900  # 15 minutes
     REFRESH_TOKEN_EXPIRATION = 2592000  # 30 days
-    NUMBER_OF_CONTENT_IN_ANN = 100 # UPDATE THIS WHEN DEVELOPING ANN
+    NUMBER_OF_CONTENT_IN_ANN = 1000 # UPDATE THIS WHEN DEVELOPING ANN
```
Comment (kennethgoodman) [inline services/backend/src/config.py:11]: please remove this change

File: services/backend/src/data_structures/approximate_nearest_neighbor/two_tower_ann.py [hunk 13]
```diff
@@ -118,7 +117,6 @@ def get_ANN_recommendations_from_content(content_id, team, K):
         content_df = pd.DataFrame(user_engagements_for_content)
 
         content_embedding = team_wrappers[team].generate_content_embeddings(content_df)
-
```
Comment (kennethgoodman) [inline services/backend/src/data_structures/approximate_nearest_neighbor/two_tower_ann.py:121]: please remove this change

File: services/backend/src/recommendation_system/ml_models/echo/scaler_state.pkl
Comment (kennethgoodman) [inline services/backend/src/recommendation_system/ml_models/echo/scaler_state.pkl:1]: is this used? if not, remove this

File: services/backend/src/recommendation_system/ml_models/echo/encoder_state.pkl
Comment (kennethgoodman) [inline services/backend/src/recommendation_system/ml_models/echo/encoder_state.pkl:1]: is this used? if not remove

**Review by kennethgoodman — COMMENTED** [review]
please address the comments I put in your files

File: services/backend/df_user.csv
Comment (kennethgoodman) [inline services/backend/df_user.csv:1]: please remove if not used

**Comment by JYS-99** [comment]
Hi Prof. Goodman,

is there any other issue with our codes? Thanks!

File: services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py [hunk 14]
```diff
@@ -17,25 +19,69 @@ def __new__(cls):
...
-        # Fetch the list of similar users for a given user_id from the map.
+
         return self.user_similarity_map.get(user_id, [])
 
     def recommend_items(self, user_id, num_recommendations=10):
-        # For a given user, fetch the list of similar users.
-        # Recommend items engaged by those users, which the given user hasn't seen.
-        pass
+
+        self.compute_similarity()
```
Comment (kennethgoodman) [inline services/backend/src/data_structures/user_based_recommender/echo/UserBasedRecommender.py:65]: shouldn't need to this is every time, its already done in __new__

**Review by kennethgoodman — COMMENTED** [review]
there are some comments to address

**Comment by kennethgoodman** [comment]
Please look at the "files changed" tab

**Review by kennethgoodman — APPROVED** [review]

**Pull request merged**